### DISCUSSION TITLE
dynamic DAGs get deactivated and removed from DAG bag

### DISCUSSION THREAD

**Discussion by rdjouder** [apache/airflow#27133] [Q&A]
Hi,
I am upgrading from 2.1.4 to 2.4.1. My DAGs are generated automatically by a single script "dags/dynamic_dag_generator.py". I have no other DAG python file is the dags folder.
To keep the DAG active it was (as of 2.1.4) enough to keep each instance/object in a global variable. As of 2.4.1 I can see that the manager is deactivating my DAGs after their first run. The problem is that they have daily scheduling. So seeing them off the dagbag is really worrying.

{manager.py:520} INFO - DAG dag_name is missing and will be deactivated.
{manager.py:530} INFO - Deactivated 1 DAGs which are no longer present in file.

 I have tried to use @dag and @task decorators for DAG generation, my DAGs are still deactivated by manager.
Do I have to generate a dag py (in dags folder) script for each DAG generated automatically to keep it active ?
I believe I have missed something in the documentation because this change is quite radical.
Thanks a lot for your help.

**Comment by ephraimbuddy** [comment]
Airflow now supports dynamic task mapping directly. Can you check if you can use it for your dag generation: [dynamic task mapping feature](https://airflow.apache.org/docs/apache-airflow/stable/concepts/dynamic-task-mapping.html).

**Reply by rdjouder** [reply]
Thanks a lot. Using task mapping means having one DAG which would collect all my former DAGs's tasks, is that right ?
Technically I can do this but I will lose scheduling granularity since task scheduling is based on dependencies. 
Any other option ?

**Reply by potiuk** [reply]
Can you please check @rdjouder if adding dag to global() works (see https://docs.astronomer.io/learn/dynamically-generating-dags) - generally please take a look at those examples and adjust them to follow the same patterns and see if it works.

You might also try to post parts of the code for your dags that creates the Dags so that we can help to solve your problem.

**Reply by rdjouder** [reply]
Thanks a lot for your replies. Yes I am using global(), the point is that my dynamic dag creator was not loading, creating all required DAG objects at each scheduler run. Only recently added DAGs would be "loaded" in global. So my code is incorrect, sorry for bothering you on this point. The fact is that we did this to avoid a full DB/table scan at each scheduler run.
From Airflow best practices on writing DAGs I can read:
"Specifically you should not run any database access, heavy computations and networking operations."
That's exactly what we are doing here, so we should rather use jinja tempaltes for our dynamic DAG creation. 
Can you confirm this point ?

**Reply by potiuk** [reply]
Maybe. It's impossible toi answe  your question after your description - but the best you can do is not to ask if it is fine, but do it and see if it works.
